### PR TITLE
Drop Python 3.8 for next release

### LINKED ISSUE #14284: Drop py3.8 support tracker
Coming up soon "On Apr 14, 2023 drop support for Python 3.8 (initially released on Oct 14, 2019)" (https://numpy.org/neps/nep-0029-deprecation_policy.html).

This issue is to track things that should be updated when our minimum Python version is 3.9.

(Possibly relevant) New features in Python:
- union operators for dict (I think pyupgrade will fix most of these)
- relaxed decorator grammar
- string methods to remove prefixes and suffixes.
- `os.readlink` can be replaced by `pathlib.Path.readlink`

### PR DESCRIPTION
**Pull request by nstarman**
Fixes #14284 

### After merge

- [ ] Update branch protection rules for `main`.
- [ ] Open follow-up issue about https://github.com/astropy/astropy/pull/14286#discussion_r1170415098

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label. Codestyle issues can be fixed by the [bot](https://docs.astropy.org/en/latest/development/workflow/development_workflow.html#pre-commit).
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Comment by nstarman** [comment]
Hm. Since I bumped all the python version in all the tests, GitHubs ``(Required)`` branch protections need to be updated.

**Review by mhvk — COMMENTED** [review]
Have only glanced over this, but it looks good - nice to get rid of those typing work-arounds!
(Aside, I mentioned numpy 1.20 in #14284, but good to do that in a separate PR.)

**Comment by nstarman** [comment]
I think the 32-bit and parallel py wasn't bumped.
Also, do we need to bump the minimum matplotlib version?

**Comment by nstarman** [comment]
The ``CI / tests / Python 3.9 with oldest supported version of all dependencies (pull_request) `` appears to be failing because of matplotlb.

**Comment by nstarman** [comment]
The py3.11 test failure is caused by a failure in ``astropy/utils/data.py``. Something about downloading a file over FTP....

**Comment by pllim** [comment]
Same comment as https://github.com/astropy/astropy/issues/14284#issuecomment-1397054127

**Comment by pllim** [comment]
At the infrastructure tag-up on 2023-02-28, release managers had given a go to drop Python 3.8 for astropy v5.3. It is also agreed that we should keep introducing Python 3.8+ specific stuff to a minimum for now to ensure we can still backport to 5.0.x and 5.2.x without much pain.

@nstarman , there are conflicts now that need to be resolved. Are you still interested to pursue this? Thanks for your patience!

**Comment by nstarman** [comment]
I can rebase and fix. I'll keep pyupgrade targeting py3.8.

**Comment by nstarman** [comment]
Much of the py3.8 to 3.9 transition is contained in https://github.com/astropy/astropy/pull/14286/commits/76879800b0452fb1d9d5b5732c3d162041385f35. I can pair stuff down. Adding ``from __future__ import annotations`` will work around the `typing` upgrades, but the simplifications in `units` can't be worked around.

**Comment by nstarman** [comment]
Ok. FTP error still happening.
https://github.com/astropy/astropy/actions/runs/4300114380/jobs/7496008367#step:6:1917
@dhomeier, you most recently edited that file. Do you have any insights? Thanks!

**Comment by mhvk** [comment]
@nstarman - I think it is completely fair game to remove work-arounds for older python versions if we change our minimum supported version. The hesitation some of us have is for broadly changing things like `dict.update(...)` to `dict |= ...` just because the language now allows it. For those, the net benefit is just not obvious. In contrast, for units to just be able to use typing without work-arounds is a very clear benefit (as is being able to use `list[...]` instead of `List` which always felt incredibly ugly and hacky) - so I'm all for those!

File: astropy/cosmology/core.py
Comment (mhvk) [inline astropy/cosmology/core.py:24]: I'm confused, why is this not imported unconditionally? Why the need for `TYPE_CHECKING`?

File: astropy/modeling/bounding_box.py
Comment (mhvk) [inline astropy/modeling/bounding_box.py:4]: I thought the license comment is always above the module docstring (it is elsewhere)

**Review by mhvk — COMMENTED** [review]
Looks good to me, modulo a few small comments/questions.

File: astropy/cosmology/core.py
Comment (nstarman) [reply astropy/cosmology/core.py:24]: If an import is used only in a type-checking context, without a runtime requirement, then we don't need to import it during runtime. I advocated to get this check into Ruff because it helps to minimize runtime imports and import cycles.

Comment (nstarman) [reply astropy/cosmology/core.py:24]: For example, consider type checking a function that can take a SciPy object, but we don't have SciPy as a runtime dependency. Separating runtime from non-runtime imports would allow us to both type hint the function and not have the runtime dependency. Ruff now sorts our imports and (can or does, I forget which setting) remove unused imports. It also tells us which imports are used in what context. Hopefully autofix is coming soon, so type-checking-only imports will be auto-sorted into `if TYPE_CHECKING` blocks.

Comment (mhvk) [reply astropy/cosmology/core.py:24]: OK, thanks, that makes sense. I was surprised partially since now the annotation refers to an object that is not necessarily imported. But perhaps this is where the lazy evaluation of annotations that I vaguely recall being discussed comes in?

Anyway, if it works, that's great!

Comment (nstarman) [reply astropy/cosmology/core.py:24]: Yes, the lazy evaluation of annotations is a complicated landscape. 
https://peps.python.org/pep-0563/ vs https://peps.python.org/pep-0649/. When it's eventually settles, that'll be great.

**Comment by mhvk** [comment]
I had a quick scan and I think this is all OK! I like that typing no longer needs the ugly work-arounds. Since I already essentially approved it above, it is probably better for someone else to check (if you think it is now ready for review; it certainly seems so!)

**Comment by nstarman** [comment]
All tests passing except ``Python 3.9 with oldest supported version of all dependencies``. I think we might need to bump the minimum matplotlib and / or scipy.

**Comment by pllim** [comment]
I think `Bad Gateway` error in dev job is transient.

**Comment by pllim** [comment]
Looks like I have to update the branch protection rules for `main` after this is merged.

**Comment by pllim** [comment]
I pinged @eteq about https://affiliated.astropy.org/ being down.

File: .github/workflows/ci_cron_weekly.yml
Comment (pllim) [inline .github/workflows/ci_cron_weekly.yml:103]: Why do we put Python version in the name? It makes no sense. 😆

Comment (pllim) [reply .github/workflows/ci_cron_weekly.yml:103]: It does not control the Python version. Even on `main`, it already picks up Python 3.10 even though we say Python 3.9 in the name: https://github.com/astropy/astropy/actions/runs/4593442983/jobs/8111372319 

🤷

File: examples/README.rst
Comment (pllim) [inline examples/README.rst:11]: Maybe we should just link to the final destination. It was redirected anyway.

```suggestion
`affiliated packages <https://www.astropy.org/affiliated/index.html>`_) belong at
```

File: pyproject.toml
Comment (pllim) [inline pyproject.toml:529]: Shouldn't the `mpl322` be `mpl33x` now with the matplotlib minversion bump?

Comment (pllim) [reply pyproject.toml:529]: Might need to generate new truths for https://github.com/astropy/astropy-figure-tests too.

Comment (pllim) [reply pyproject.toml:529]: Maybe @astrofrog or @Cadair can confirm of such procedures when we bump matplotlib minversion.

I think the CircleCI config also needs updating.

**Review by pllim — CHANGES_REQUESTED** [review]
I think bumping matplotlib needs more work.

Probably don't need what's new, but definitely need a change log about the minversion bumps here.

HTTP -> HTTPS seems unrelated but doesn't hurt, so ok to keep.

**Comment by eerovaher** [comment]
We have at least one doctest that is currently being skipped because it wouldn't work with Python 3.8, but if support for that is being dropped then the test shouldn't be skipped anymore: https://github.com/astropy/astropy/blob/140051b0691c8fd1cc4bc4860c59803a96485d95/docs/coordinates/index.rst#L426-L440

File: pyproject.toml
Comment (nstarman) [reply pyproject.toml:529]: Yes, help with that would be appreciated!

Comment (nstarman) [reply pyproject.toml:529]: @pllim, I added the Changelog you suggested.

**Comment by nstarman** [comment]
> We have at least one doctest that is currently being skipped because it wouldn't work with Python 3.8, but if support for that is being dropped then the test shouldn't be skipped anymore:

👍 I checked and found some more.

File: pyproject.toml
Comment (nstarman) [reply pyproject.toml:529]: > I think bumping matplotlib needs more work.

@pllim , other than this thread, what else about bumping matplotlib needs doing?

**Comment by nstarman** [comment]
I think the link check failure is unrelated. The http -> https commit happened when I was looking over the logs. There are a lot more redirects we should eventually address.

**Review by WilliamJamieson — APPROVED** [review]
The modeling changes look good to me. I looked over the rest of the changes and those seem okay to me as well. I think others who know those parts of astropy better should also approve before this gets merged.

**Comment by pllim** [comment]
There is a conflict now. Can you pls rebase?

As for matplotlib minversion, I'll have to run some grep when I have time to see if you have caught them all.

File: docs/coordinates/index.rst
Comment (eerovaher) [inline docs/coordinates/index.rst:434]: If `tz_name` was defined in a setup block then this wouldn't have to be a remote test.

**Review by eerovaher — COMMENTED** [review]
The following lines should be deleted because the code they are referring to will be tested regularly. https://github.com/astropy/astropy/blob/b483f46ccfb6e81f5b84e91ece18b8014e7b50a7/docs/coordinates/index.rst#L436-L438

File: docs/coordinates/index.rst
Comment (nstarman) [reply docs/coordinates/index.rst]: Done!

**Comment by pllim** [comment]
Why is this failing for some jobs?

```
>>> tz = ZoneInfo(tz_name)
UNEXPECTED EXCEPTION: ZoneInfoNotFoundError('No time zone found with key America/Phoenix')
```

**Comment by nstarman** [comment]
Trying to resolve https://github.com/astropy/astropy/pull/14286#discussion_r1164537029

File: docs/coordinates/index.rst
Comment (nstarman) [reply docs/coordinates/index.rst]: @eerovaher, errors on the cron jobs only.

**Comment by pllim** [comment]
We should really wrap this up and merge early next week if we still want to drop Python 3.8 for v5.3. It won't do to merge this one last-last minute. If a change broke CI and is unrelated to dropping Python 3.8, open a follow-up issue for it and revert that change. Thanks!

File: docs/coordinates/index.rst
Comment (nstarman) [reply docs/coordinates/index.rst:434]: We can fix this in a followup.

**Comment by nstarman** [comment]
Done?! 🤞

**Comment by pllim** [comment]
Still not sure about the mpl bump, so I'll re-review in a couple of days. Thanks for your patience!

File: astropy/coordinates/tests/test_velocity_corrs.py
Comment (eerovaher) [inline astropy/coordinates/tests/test_velocity_corrs.py:1]: This pull request shouldn't touch this file at all.

File: docs/coordinates/index.rst
Comment (eerovaher) [reply docs/coordinates/index.rst:434]: I'd prefer this example to be included in the non-remote tests, but with the feature freeze approaching it is indeed best to leave that for a followup. Having it included in the remote tests is an improvement already.

**Review by eerovaher — CHANGES_REQUESTED** [review]
I agree that the doctest involving timezones doesn't have to be sorted out in this pull request, but it looks like something's gone wrong during merge conflict resolution and the changes to `astropy/coordinates/tests` should be undone.

File: docs/coordinates/index.rst
Comment (pllim) [reply docs/coordinates/index.rst:434]: The `of_address` always needs remote data access, so I don't understand your original comment above, but this is unrelated to dropping Python 3.8, so please open a follow-up issue now so we don't forget about it. Thanks!

File: setup.cfg
Comment (pllim) [inline setup.cfg:99]: Where did this come from?

Comment (pllim) [reply setup.cfg:99]: Ah, so this is the cost of enabling that coordinates test above. Hmm... I guess it is ok for now. But if it causes too much maintenance headache in the future, we can revisit this.

Comment (nstarman) [reply setup.cfg:99]: Will do!

Comment (pllim) [reply setup.cfg:99]: BTW, in case you missed the Slack msg, I will fold this into my edit and I plan to push to this branch directly, so pls don't touch it or you might lose your changes. FYI. Thanks!

File: docs/coordinates/index.rst
Comment (eerovaher) [reply docs/coordinates/index.rst:434]: There's two related examples here close to each other. The first example shows how to get the name of a timezone and that can be left as a remote test. The second example demonstrates using a `ZoneInfo` object created from the timezone name and can be made a non-remote test if `tz_name` is defined in a setup block.

**Review by eerovaher — APPROVED** [review]

**Comment by pllim** [comment]
linkcheck failed at https://ieeexplore.ieee.org/document/5254933 but it is not a broken link, so maybe the server temporary blacklisted our CI. We can ignore the failure for now.

A bunch of the "required" tests that won't run will block normal merge, so I will use my admin power to merge when CI passes. I need to update the branch protection rules with new CI names but only after merging this. 🐔 🥚 

Thanks everyone, especially to @nstarman for your patience!

**Comment by pllim** [comment]
Alas, the protection rules will change and impact all the other PRs wanting to make it to feature freeze. I will update the rules and send out a memo to astropy-dev.

File: docs/coordinates/index.rst
Comment (pllim) [reply docs/coordinates/index.rst:434]: I opened follow-up issue at https://github.com/astropy/astropy/issues/14646

**Pull request merged**
